src/Features/LanguageServer/Protocol/Handler/RequestContext.cs

```diff
@@ -89,33 +90,37 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
             Dictionary<Workspace, (Solution workspaceSolution, Solution lspSolution)>? solutionCache,
             IDocumentChangeTracker? documentChangeTracker,
             ImmutableArray<string> supportedLanguages,
-            IGlobalOptionService globalOptions,
-            out Workspace workspace)
+            IGlobalOptionService globalOptions)
         {
             // Go through each registered workspace, find the solution that contains the document that
             // this request is for, and then updates it based on the state of the world as we know it, based on the
             // text content in the document change tracker.
 
-            // Assume the first workspace registered is the main one
-            var workspaceSolution = lspWorkspaceRegistrationService.GetAllRegistrations().First().CurrentSolution;
-            Document? document = null;
-
-            // If we were given a document, find it in whichever workspace it exists in
-            if (textDocument is null)
+            Document? document;
+            Solution? workspaceSolution;
+            if (textDocument is not null)
             {
-                logger.TraceInformation("Request contained no text document identifier");
+                // we were given a request associated with a document.  Find the corresponding roslyn
+                // document for this.  If we can't, we cannot proceed.
+                document = FindDocument(logger, telemetryLogger, lspWorkspaceRegistrationService, lspMiscellaneousFilesWorkspace, textDocument, clientName);
+                if (document == null)
+                {
+                    logger.TraceError($"Could not find document: {textDocument.Uri}");
+                    return default;
```

Comment:
This change in behaviour scares me. Is it necessary? Based on the description I would expect this would just call `TryGetHostWorkspace`?

Reply:
to me it seems sensible.  if we were asked for a particular document uri, and we can't find it... then we should not be able to proceed.  what woudl it mean to be asked to answer something about a doc we don't know about?

Reply:
I guess my thinking is inline with your comment in `ExecuteAsync` "it may be necessary to allow individual handlers to decide what to return in those cases.". Essentially thats what the previous code was doing - it was up to an individual handler to decide if a null document was important to them.

I think it's certainly within the realms of possibilities that there might be a handler that always answers the same way regardless of what document its called for, but I certainly can't think of any examples within Roslyn.

Reply:
manually checked the lsp spec for everything.  all the commands that could get a doc are allowed to return null.